### PR TITLE
[Relay] Reference

### PR DESCRIPTION
**Pull request by MarisaKirisame**
See #2483

### CODE REVIEW

**Comment by MarisaKirisame** [comment]
@ZihengJiang @junrushao1994 @zhiics can you guys review this? Thx!

File: src/relay/pass/type_infer.cc [hunk 1]
```diff
@@ -390,6 +390,23 @@ class TypeInferencer : private ExprFunctor<Type(const Expr&)> {
...
   }
+
+  Type VisitExpr_(const RefNewNode* op) final {
+    return RefTypeNode::make(GetType(op->value));
+  }
+
+  Type VisitExpr_(const RefReadNode* op) final {
+    Type it = IncompleteTypeNode::make(TypeVarNode::Kind::kType);
+    this->Unify(GetType(op->ref), RefTypeNode::make(it), op->span);
+    return it;
```
Comment (slyubomirsky) [inline src/relay/pass/type_infer.cc]: Why unify with a fresh incomplete type? You could just return `GetType(op->ref)`

Comment (MarisaKirisame) [reply src/relay/pass/type_infer.cc]: this propagate the constraint backward (it unify op->ref with a reftype

Comment (slyubomirsky) [reply src/relay/pass/type_infer.cc]: Ah, I misread the second line. Thanks for clarifying

**Comment by slyubomirsky** [comment]
Unit tests for new alpha equality and type inference cases would be helpful for pinpointing bugs/regression.

**Comment by MarisaKirisame** [comment]
@tqchen the RFC is the same RFC (#2237 ).

File: src/relay/ir/expr.cc [hunk 2]
```diff
@@ -271,13 +271,58 @@ TVM_STATIC_IR_FUNCTOR_REGISTER(IRPrinter, vtable)
...
 });
 
+RefNew RefNewNode::make(Expr value) {
+  NodePtr<RefNewNode> n = make_node<RefNewNode>();
+  n->value = std::move(value);
+  return RefNew(n);
+}
 
-TVM_REGISTER_API("relay._expr.TempExprRealize")
+TVM_REGISTER_API("relay._make.RefNew").set_body([](TVMArgs args, TVMRetValue* ret) {
```
Comment (ZihengJiang) [inline src/relay/ir/expr.cc]: another line for `.set_body`

File: python/tvm/relay/expr.py [hunk 3]
```diff
@@ -327,6 +327,46 @@ def __init__(self, tuple_value, index):
...
+    """
+    def __init__(self, ref):
+        self.__init_handle_by_constructor__(_make.RefRead, ref)
+
+
+@register_relay_node
+class RefWrite(Expr):
+    """
+    Update the value inside the reference.
+    The whole expression will evaluate to an empty tuple.
```
Comment (ZihengJiang) [inline python/tvm/relay/expr.py]: Why choose empty tuple

**Comment by zhiics** [comment]
I only have some minor comments about the PR.

But I do have one concern about adding AST nodes/types because it would affect most relay passes. I see that some passes are modified correspondingly, but how about others such as memory planning and forward rewriting, etc?

File: src/relay/backend/interpreter.cc [hunk 4]
```diff
@@ -432,6 +449,31 @@ class Interpreter :
     }
   }
 
+  Value VisitExpr_(const RefWriteNode* op) final {
+    Value r = Eval(op->ref);
+    if (const RefValueNode* rv = r.as<RefValueNode>()) {
+      rv->value = Eval(op->value);
+      return TupleValueNode::make({});
```
[the remark this answers is not in the document]
Comment (MarisaKirisame) [reply src/relay/backend/interpreter.cc]: yes. there is nothing needed to be return, so an empty tuple will do the job.

File: include/tvm/relay/expr.h [hunk 5]
```diff
@@ -428,12 +428,78 @@ class TupleGetItemNode : public ExprNode {
...
 
-  static constexpr const char * _type_key = "relay.TupleGetItem";
+  static constexpr const char* _type_key = "relay.TupleGetItem";
   TVM_DECLARE_NODE_TYPE_INFO(TupleGetItemNode, ExprNode);
 };
 
 RELAY_DEFINE_NODE_REF(TupleGetItem, TupleGetItemNode, Expr);
 
+/*! \brief Create a new Reference out of initial value. */
+class RefNew;
```
[the remark this answers is not in the document]
Comment (MarisaKirisame) [reply include/tvm/relay/expr.h]: sgtm

**Comment by MarisaKirisame** [comment]
@zhiics how do you do memory planning on control flow? i imagine planning for reference will be similar.
irregardless, i dont think it is much of a big deal - it is only used for higher order ad, and will be removed by a pass that try to fuse reference. if it cannot be fused it is pytorch style nlp model, and since tvm does not support them by now any improvement is good.

**Pull request merged**